Messaging.md

```diff
@@ -363,92 +371,196 @@ Witnesses should be sorted by the `input_serial_id` sent in `funding_input` defi
-
 #### `oracle_event`
 
-1. type: 55330
 1. data:
-   * [`u16`:`nb_nonces`]
-   * [`nb_nonces*x_point`:`oracle_nonces`]
-   * [`u32`:`event_maturity_epoch`]
+   * [`oracle_event_timestamp`:`timestamp`]
    * [`event_descriptor`:`event_descriptor`]
    * [`string`:`event_id`]
 
+### The `oracle_metadata` Type
+
+This type contains static oracle information and can be used to import trusted oracles into wallets.
+This information should be saved by DLC nodes.
+
+#### `oracle_metadata`
+
+1. data:
+   * [`x_point`:`announcement_public_key`]
+   * [`string`:`oracle_name`]
+   * [`string`:`oracle_description`]
+   * [`u32`:`timestamp`]
+   * [`oracle_schemes`:`oracle_schemes`]
+   * [`signature`:`oracle_metadata_signature`]
+
+where the `oracle_metadata_signature` is a Schnorr signature by the `announcement_public_key` using the tag `oraclemetadata/v0` of the sha256 hash of the serialized `oracle_metadata` message omitting the `announcement_public_key` and of course the the `oracle_metadata_signature`.
+
+#### Rationale
+
+Separate public keys are required for announcement and attestation because the attestation scheme
+used in the current DLC spec reduces the security of non-attestation Schnorr signatures issued by the
+attestation keys and hence a separate announcement key is required.
+
+The `oracle_metadata_signature` is used to link all data in this message to the `announcement_public_key`.
+
+#### Requirements
+
+A node
+
+* SHOULD save this information in persistent storage.
+* MUST reject if `announcement_public_key == attestation_public_key`.
+
+### The `oracle_schemes` Type
+
+In order to enable oracles to support multiple attestation schemes, and clients to chose their preferred one, the `oracle_schemes` type can contain multiple attestations schemes.
+Each scheme is defined as a TLV so that clients can easily ignore schemes that they don't know about.
+Note that at the moment the only defined scheme is the `schnorr_scheme` that uses Schnorr signatures.
+
+#### `oracle_schemes`
+
+1. data:
+   * [`bigsize`:`nb_schemes`]
+   * [`nb_schemes*scheme`:`attestation_schemes`]
+
+#### `schnorr_scheme`
+
+1. type: 0
+1. data:
+   * [`x_point`:`attestation_public_key`]
+   * [`bigsize`:`nb_nonces`]
+   * [`nb_nonces*x_point`:`oracle_nonces`]
+   * [`signature:`proof_of_knowledge`]
+
+where `proof_of_knowledge` ensures that the oracle knows the pre-image of the `attestation_public_key` and all `oracle_nonce`.
+It is generated following bip-musig2 [signing](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#signing) and [signature aggregation](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#partial-signature-aggregation) algorithms, using the attestation key as the first key and the nonces as the following keys, in the same order as provided in the data structure, and the message `dlcoraclepok`.
```

Comment:
So does this block implementation on this until all libraries support musig2?

Reply:
I guess it's up for us to decide. The alternatives to musig is to do separate proofs of knowledge, which is fine but we'll end up with rather big announcements for numerical events, with a signature per nonce.

Note it's mainly about binding to the musig2 implementation, once you have the bindings it's quite little work (you can see my implementation [here](https://github.com/p2pderivatives/rust-dlc/commit/3bf2d7c6904ba7309154c4c30104ab55f525a063)

Reply:
Do you have a musig2 implementation available to you?

Also does @matthewjablack have a musig2 implementation available to him? 

We are _just_ implementing it: https://github.com/bitcoin-s/bitcoin-s/pulls

Reply:
Also this brings up stability questions of the current musig2 spec. For instance, I know Schnorr was deployed on the bitcoin cash blockchain years before schnorr got deployed into Bitcoin. 

I think things like `XOnlyPubKey`s weren't included in the schnorr spec, thus leading to incompatibilities between the two implementations. 

We could end up in a similar situation if we don't wait for the specification to mature. I guess lnd is deploying musig2 (although not incorporating it in the LN spec, so its easier to change for now!)

https://docs.bitcoincashnode.org/doc/bips/

![Screenshot from 2022-06-27 07-37-37](https://user-images.githubusercontent.com/3514957/175943545-567f8751-86c1-4671-be87-f1eec4643275.png)

TLDR, how stable is musig2?

Reply:
Taproot hasn't been added to bitcoinjs-lib let alone musig2

If we wanted to get access to musig2 we'd likely need to get it from secp256k1-zkp and update cfd/cfd-core

This is a little annoying since it would mean validation for Oracle Announcements would require cfd-dlc-js and couldn't be done in native JS as it's done currently https://github.com/AtomicFinance/node-dlc/blob/master/packages/messaging/lib/messages/OracleAnnouncementV0.ts#L62

Reply:
@Christewart
> Do you have a musig2 implementation available to you?

Yes, WIP but it's there: https://github.com/ElementsProject/rust-secp256k1-zkp/pull/48

> TLDR, how stable is musig2?

I don't know, maybe @LLFourn does?

@matthewjablack 

> This is a little annoying since it would mean validation for Oracle Announcements would require cfd-dlc-js and couldn't be done in native JS as it's done currently https://github.com/AtomicFinance/node-dlc/blob/master/packages/messaging/lib/messages/OracleAnnouncementV0.ts#L62

Two possible solutions would be use a wasm module generated from either cfd-core or rust-secp256k1-zkp (I think it could work, but honestly haven't tried yet), or if you have access to EC operations in JS implement the key aggregation algorithm which is all you need on client side (not sure it'd be a good idea tbh, and it might take some work).

We can probably discuss at the next meeting whether we want to do that or not. I think it's a bit nicer as it gives more compact and constant size announcements, but if people prefer to have separate signatures we can also go that way.

Reply:
whatever happens with DLCs going forward I expect MuSig2 will be part of it so I think relying on the MuSig key aggregation function is not too much to ask to simplify things a bit. It should be stable soon.